### PR TITLE
Introduce validations and variable order

### PR DESCRIPTION
**Pull request by zackproser**
## Description

This PR is intended to be merged to master in lieu of #105, and it is ready for review!

These changes introduce two new features to boilerplate: 
* The ability to define a set of validations that user submissions must include when prompting a human for a variable value
* The ability to define the order in which a set of given boilerplate variables 

### Validations 

Validations and prompt order are defined in your `boilerplate.yml` like so: 

```yaml
variables:
  - name: CompanyName
    order: 0
    description: |
      Enter the name of your organization. This will be used to generate a prefix name for every resource.

      Example: gruntwork
    default: ""
    type: string
    validations:
      - required
      - length-5-22
      - alphanumeric
```
In this example variable definition, we're defining a prompt order of `0`, meaning this variable will always be shown to users first. We also define three validations that any user submission for this variable *must* satisfy. Any user-submissions that do not pass validation will be rejected at runtime with a helpful error message, like so: 

bp-validation

## TODOs

Read the [Gruntwork contribution guidelines](https://gruntwork.notion.site/Gruntwork-Coding-Methodology-02fdcd6e4b004e818553684760bf691e).

- [x] Update the docs.
- [x] Run the relevant tests successfully, including pre-commit checks.
- [x] Ensure any 3rd party code adheres with our [license policy](https://www.notion.so/gruntwork/Gruntwork-licenses-and-open-source-usage-policy-f7dece1f780341c7b69c1763f22b1378) or delete this line if its not applicable.
- [x] Include release notes. If this PR is backward incompatible, include a migration guide.

## Release Notes (draft)

Implement variable validations and relative ordering.

### CODE REVIEW

File: config/get_variables.go [hunk 1]
```diff
@@ -54,7 +59,40 @@ func GetVariables(opts *options.BoilerplateOptions, boilerplateConfig, rootBoile
...
+	// process for our Ref Arch customers
+
+	// Create a slice of KeyOrderPairs, which we'll be able to sort according to order value
+	keyAndOrderPairs := []KeyAndOrderPair{}
+
+	// Pair the keys for each variable to its user-defined presentation order
+	for key, variable := range variablesInConfig {
+		kop := KeyAndOrderPair{
+			Key:   key,
+			Order: variable.Order(),
```
Comment (brikis98) [inline config/get_variables.go:79]: If the user doesn't specify `order`, or they define it for some inputs but not all, is the default is 0? If so, does that mean vars without `order` go before the ones with `order`?

File: .gitignore [hunk 2]
```diff
@@ -2,4 +2,5 @@
 vendor
 .terragrunt
 .DS_Store
-*.iml
\ No newline at end of file
+*.iml
+boilerplate
```
Comment (brikis98) [inline .gitignore:6]: NIT: might want to add a comment on what this is (the binary).

File: README.md [hunk 3]
```diff
@@ -125,6 +125,8 @@ You can find older versions on the [Releases Page](https://github.com/gruntwork-
    which gives you the ability to do formatting, conditionals, loops, and call out to Go functions. It also includes
    helpers for common tasks such as loading the contents of another file.
 1. **Variable types**: Boilerplate variables support types, so you have first-class support for strings, ints, bools,
+1. **Validations**: Boilerplate provides a set of validations for a given variable that user input must satisfy.
+1. **Variable presentation order**: Boilerplate allows you to define the relative presentation order of a set of variables.
    lists, maps, and enums.
```
Comment (brikis98) [inline README.md:130]: Looks like this list of types got cut off from the `**variable types**` entry above and ended up on the next line.

File: config/get_variables.go [hunk 4]
```diff
@@ -54,7 +59,40 @@ func GetVariables(opts *options.BoilerplateOptions, boilerplateConfig, rootBoile
...
+	// N.B. this syntax of sort.Slice requires Go 1.18 or above!
+	sort.Slice(keyAndOrderPairs[:], func(i, j int) bool {
+		return keyAndOrderPairs[i].Order < keyAndOrderPairs[j].Order
+	})
+
+	// Now, instead of just iterating through the map keys naively,
+	// iterate through the slice of KeyOrderPairs, which are sorted by order
+	// which means that in each iteration of the loop, we can fetch the next variable
+	// by looking up its key in the original config-provided variables map
+	for _, keyOrderPair := range keyAndOrderPairs {
```
Comment (brikis98) [inline config/get_variables.go:94]: The ordering code looks good, but do we have any automated test to check that it actually works? Can be added in a follow-up PR too.

File: config/get_variables.go [hunk 5]
```diff
@@ -193,15 +233,92 @@ func getVariableFromUser(variable variables.Variable, opts *options.BoilerplateO
 		}
 	}
 
+	var valueToValidate interface{}
+	if value == "" {
```
Comment (brikis98) [inline config/get_variables.go:237]: Is there any way for the user to enter an empty string as their desired value?

File: config/get_variables.go [hunk 6]
```diff
@@ -144,27 +194,17 @@ func getVariableFromVars(variable variables.Variable, opts *options.BoilerplateO
...
-	}
-
-	helpText := []string{
-		fmt.Sprintf("type: %s", variable.Type()),
-	}
-
-	if variable.ExampleValue() != "" {
-		helpText = append(helpText, fmt.Sprintf("example value: %s", variable.ExampleValue()))
-	}
+func getVariableFromUser(variable variables.Variable, opts *options.BoilerplateOptions, invalidEntries variables.InvalidEntries) (interface{}, error) {
```
Comment (brikis98) [inline config/get_variables.go:197]: NIT: consider refactoring this into several smaller methods. It has gotten quite large and is doing a lot of totally separate things in a row, with lots of mutation in between, so it's getting harder to follow. Can be done in a follow-up PR.

File: config/get_variables.go [hunk 7]
```diff
@@ -193,15 +233,92 @@ func getVariableFromUser(variable variables.Variable, opts *options.BoilerplateO
...
+	if value == "" {
+		valueToValidate = variable.Default()
+	} else {
+		valueToValidate = value
+	}
+	// If any of the variable's validation rules are not satisfied by the user's submission,
+	// store the validation errors in a map. We'll then recursively call get_variable_from_user
+	// again, this time passing in the validation errors map, so that we can render to the terminal
+	// the exact issues with each submission
+	m := make(map[string]bool)
```
Comment (brikis98) [inline config/get_variables.go:246]: NIT: should this be a list of structs, where each struct contains a validation and whether it passed or not? That way, the list of validations will be in a consistent order each time we display it, whereas with a map, the order will change. Can be done in another PR.

File: config/get_variables.go [hunk 8]
```diff
@@ -193,15 +233,92 @@ func getVariableFromUser(variable variables.Variable, opts *options.BoilerplateO
...
 
+	// Clear the terminal of all previous text for legibility
+	util.ClearTerminal()
+
+	if variable.Type() == variables.String {
+		_, intErr := strconv.Atoi(value)
+		if intErr == nil {
+			value = fmt.Sprintf(`"%s"`, value)
+		}
+	}
```
Comment (brikis98) [inline config/get_variables.go:284]: Why is this here?

File: config/get_variables.go [hunk 9]
```diff
@@ -193,15 +233,92 @@ func getVariableFromUser(variable variables.Variable, opts *options.BoilerplateO
...
+
 	return variables.ParseYamlString(value)
 }
 
+// RenderValidationErrors displays in user-legible format the exact validation errors
+// that the user's last submission generated
+func renderValidationErrors(val interface{}, m map[string]bool) {
```
Comment (brikis98) [inline config/get_variables.go:291]: NIT: avoid single-letter variable names in most cases. https://docs.gruntwork.io/guides/style/golang-style-guide/#naming-things. Can be done in a separate PR.

File: config/get_variables.go [hunk 10]
```diff
@@ -193,15 +233,92 @@ func getVariableFromUser(variable variables.Variable, opts *options.BoilerplateO
...
+
+func renderVariablePrompts(variable variables.Variable, invalidEntries variables.InvalidEntries) {
+	pterm.Println(pterm.Green(variable.FullName()))
+
+	if variable.Description() != "" {
+		pterm.Println(pterm.Yellow(variable.Description()))
+	}
+
+	if len(invalidEntries.Issues) > 0 {
+		renderValidationErrors(invalidEntries.Issues[0].Value, invalidEntries.Issues[0].ValidationMap)
```
Comment (brikis98) [inline config/get_variables.go:311]: Do we only show the first issue intentionally?

File: variables/types.go [hunk 11]
```diff
@@ -38,6 +38,15 @@ func (boilerplateType BoilerplateType) String() string {
 	return string(boilerplateType)
 }
 
+type InvalidEntries struct {
```
Comment (brikis98) [inline variables/types.go:41]: NIT: add comments explaining intended usage for this struct and the one below. Could be separate PR.

File: variables/yaml_helpers.go [hunk 12]
```diff
@@ -175,6 +178,130 @@ func unmarshalOptionsField(fields map[string]interface{}, context string, variab
...
+
+	return validationRules, nil
+}
+
+// Given a list of validations read from a Boilerplate YAML config file of the format:
+//
+// This method looks up the validations specified in the map and applies them to the specified fields so that users prompted for input
+// get real-time feedback on the validity of their entries
+func unmarshalValidationsField(fields map[string]interface{}, context string, variableType BoilerplateType) ([]CustomValidationRule, error) {
+	validations, _ := fields["validations"]
```
Comment (brikis98) [inline variables/yaml_helpers.go:293]: Should we be ignoring the second arg here?

File: variables/yaml_helpers.go [hunk 13]
```diff
@@ -175,6 +178,130 @@ func unmarshalOptionsField(fields map[string]interface{}, context string, variab
...
+	}
+	return validatorsToReturn
+}
+
+func (c CustomValidationRule) DescriptionText() string {
+	return c.Message
+}
+
+// parseRuleString converts the string representation of the validations field, as parsed from YAML,
+// into a slice of strings that ConvertValidationStringtoRules can easily iterate over
```
Comment (brikis98) [inline variables/yaml_helpers.go:201]: NIT: I don't have enough context to understand what this is parsing. Consider adding some examples of YAML or whatever we're parsing here to make it easier to understand. Can be in a separate PR.

File: variables/yaml_helpers.go [hunk 14]
```diff
@@ -175,6 +178,130 @@ func unmarshalOptionsField(fields map[string]interface{}, context string, variab
...
+	for _, rule := range rules {
+
+		var cvr CustomValidationRule
+
+		switch {
+		case strings.HasPrefix(rule, "length-"):
+			valString := strings.TrimLeft(rule, "length-")
+			valSlice := strings.Split(valString, "-")
+			minStr := valSlice[0]
+			maxStr := valSlice[1]
```
Comment (brikis98) [inline variables/yaml_helpers.go:225]: Should we check `len(valSlice)` is at least 2? Or exactly 2?

File: examples/for-learning-and-testing/order/boilerplate.yml [hunk 15]
```diff
@@ -0,0 +1,12 @@
+variables:
```
Comment (brikis98) [inline examples/for-learning-and-testing/order/boilerplate.yml:1]: I can't remember... Does boilerplate automatically run these examples and compare them to test fixtures as part of its automated testing?

File: config/get_variables.go (hunk 4 above)
Comment (zackproser) [reply config/get_variables.go:94]: Currently we just have the test example / fixtures that ensure using `order` in a variable definition is not going to cause a crash / panic. We do need more testing for the actual functionality in a separate go test, most likely.

File: README.md (hunk 3 above)
Comment (zackproser) [reply README.md:130]: Ah thanks!

File: .gitignore (hunk 2 above)
Comment (zackproser) [reply .gitignore:6]: Will do in follow up

File: config/get_variables.go (hunk 1 above)
Comment (zackproser) [reply config/get_variables.go:79]: Yep - that's almost certainly what will happen right now.

File: config/get_variables.go (hunk 6 above)
Comment (zackproser) [reply config/get_variables.go:197]: Agree - it's pretty gross right now. Will do in follow up

File: config/get_variables.go (hunk 7 above)
Comment (zackproser) [reply config/get_variables.go:246]: Yeah that would be nicer, indeed.

File: config/get_variables.go (hunk 8 above)
Comment (zackproser) [reply config/get_variables.go:284]: Can't remember :( - will examine in follow up

File: examples/for-learning-and-testing/order/boilerplate.yml (hunk 15 above)
Comment (zackproser) [reply examples/for-learning-and-testing/order/boilerplate.yml:1]: Yup, it does!
